src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -0,0 +1,597 @@
+        difficultyMap.get(difficulty).removeAll(selectedExercises);
+        return numberOfExercises - selectedExercises.size();
+    }
+
+    /**
+     * Computes the average confidence of all prior competencies.
+     *
+     * @param competency the competency for which the average prior confidence should be computed
+     * @param state      the current state of the recommendation (containing the mapping for prior competencies)
+     * @return the average confidence of all prior competencies
+     */
+    private static double computeCombinedPriorConfidence(Competency competency, RecommendationState state) {
+        return state.priorCompetencies.get(competency.getId()).stream().map(state.competencyIdMap::get).map(c -> c.getUserProgress().stream().findFirst())
+                .mapToDouble(competencyProgress -> {
+                    if (competencyProgress.isEmpty()) {
+                        return 0;
+                    }
+                    return competencyProgress.get().getConfidence();
+                }).sorted().average().orElse(100);
+    }
+
+    /**
+     * Predicts the number of exercises required to master the given competency based on prior performance and current progress.
+     * <p>
+     * The following formulas are used predict the number of exercises required to master the competency:
+     * <ul>
+     * <li>Mastery >= MasteryThreshold</li>
+     * <li>Mastery = 1/3 * Progress + 2/3 * Confidence</li>
+     * <li>Progress = (#CompletedLearningObjects + #ExercisesRequiredToMaster) / #LearningObjects</li>
+     * <li>Confidence = (Sum of LatestScores + #ExercisesRequiredToMaster * avg. prior Confidence) / (#LatestScores + #ExercisesRequiredToMaster)</li>
+     * </ul>
+     * The formulas are substituted and solved for #ExercisesRequiredToMaster.
+     *
+     * @param learningPath             the learning path for which the prediction should be computed
+     * @param competency               the competency for which the prediction should be computed
+     * @param priorConfidence          the average confidence of all prior competencies
+     * @param numberOfPendingExercises the number of exercises that have not been completed by the user
+     * @return the predicted number of exercises required to master the given competency
+     */
+    private int predictNumberOfExercisesRequiredToMaster(LearningPath learningPath, Competency competency, double priorConfidence, int numberOfPendingExercises) {
+        // we assume that the student may perform slightly worse that previously and dampen the prior confidence for the prediction process
+        priorConfidence *= 0.75;
+        final var scores = participantScoreService.getStudentAndTeamParticipationScoresAsDoubleStream(learningPath.getUser(), competency.getExercises()).summaryStatistics();
+        double LU = competency.getLectureUnits().size();
+        double EX = competency.getExercises().size();
+        double LO = LU + EX;
```

Comment:
What is `LO` supposed to mean? Can you use longer names?

Reply:
I guess it means learning objects but I aggree that it's really not descriptive.